nemo/collections/tts/models/hifigan.py

```diff
@@ -0,0 +1,238 @@
+            cfg = OmegaConf.create(cfg)
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        self.audio_to_melspec_precessor = instantiate(cfg.preprocessor)
+        # use a different melspec extractor because:
+        # 1. we need to pass grads
+        # 2. we need remove fmax limitation
+        self.trg_melspec_fn = instantiate(cfg.preprocessor, highfreq=None, use_grads=True)
+        self.generator = instantiate(cfg.generator)
+        self.mpd = MultiPeriodDiscriminator()
+        self.msd = MultiScaleDiscriminator()
+
+        self.sample_rate = self._cfg.preprocessor.sample_rate
+
+    def configure_optimizers(self):
+        self.optim_g = torch.optim.AdamW(
+            self.generator.parameters(), self._cfg.optim.lr, betas=[self._cfg.optim.adam_b1, self._cfg.optim.adam_b2]
+        )
+        self.optim_d = torch.optim.AdamW(
+            itertools.chain(self.msd.parameters(), self.mpd.parameters()),
+            self._cfg.optim.lr,
+            betas=[self._cfg.optim.adam_b1, self._cfg.optim.adam_b2],
+        )
+
+        self.scheduler_g = torch.optim.lr_scheduler.ExponentialLR(self.optim_g, gamma=self._cfg.optim.lr_decay,)
+        self.scheduler_d = torch.optim.lr_scheduler.ExponentialLR(self.optim_d, gamma=self._cfg.optim.lr_decay,)
+
+        return [self.optim_g, self.optim_d], [self.scheduler_g, self.scheduler_d]
+
+    @property
+    def input_types(self):
+        return {
+            "spec": NeuralType(('B', 'D', 'T'), MelSpectrogramType()),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "audio": NeuralType(('B', 'S', 'T'), AudioSignal(self.sample_rate)),
+        }
+
+    @typecheck()
+    def forward(self, *, spec):
+        """
+        Runs the generator, for inputs and outputs see input_types, and output_types
+        """
+        return self.generator(x=spec)
+
+    @typecheck(output_types={"audio": NeuralType(('B', 'T'), AudioSignal())})
+    def convert_spectrogram_to_audio(self, spec: 'torch.tensor') -> 'torch.tensor':
+        return self(spec=spec).squeeze(1)
+
+    def training_step(self, batch, batch_idx, optimizer_idx):
+        audio, audio_len = batch
+        audio_mel, dummy = self.audio_to_melspec_precessor(audio, audio_len)
+        audio = audio.unsqueeze(1)
+
+        audio_pred = self.generator(x=audio_mel)
+        audio_pred_mel, _ = self.trg_melspec_fn(audio_pred.squeeze(1), audio_len)
+
+        # train discriminator
+        self.optim_d.zero_grad()
+        mpd_score_real, mpd_score_gen, _, _ = self.mpd(y=audio, y_hat=audio_pred.detach())
+        loss_disc_mpd, _, _ = discriminator_loss(mpd_score_real, mpd_score_gen)
+        msd_score_real, msd_score_gen, _, _ = self.msd(y=audio, y_hat=audio_pred.detach())
+        loss_disc_msd, _, _ = discriminator_loss(msd_score_real, msd_score_gen)
+        loss_d = loss_disc_msd + loss_disc_mpd
+        self.manual_backward(loss_d, self.optim_d)
+        self.optim_d.step()
```

Comment:
Is there any reason, you choose to do the manual backwards as opposed to relying on lightning (like in melGAN)?

Reply:
It is more efficient that way, 1 generator forward pass for bath updates (G and D). as opposed to 1 generator forward pass for 1 generator update, and another generator pass for 1 discriminator update. this is also more in line with the optimization described in the original source code.